File: dxz/memory/block_allocator.py
class BlockAllocator:
    def __init__(self, total_blocks: int):
        self.total_blocks = total_blocks
        # free blocks' id list
        self.free_blocks: list[int] = [block_id for block_id in reversed(range(total_blocks))]

    def allocate(self, n_blocks: int) -> list[int]:
        assert n_blocks <= len(self.free_blocks), "not enough block to allocate"
        blocks: list[int] = []
        for _ in range(n_blocks):
            block_id = self.free_blocks.pop()
            blocks.append(block_id)
        return blocks
    
    def free(self, blocks: list[int]) -> bool:
        for block_id in blocks:
            assert block_id >= 0 and block_id < self.total_blocks, f"invalid block id {block_id}"
            self.free_blocks.append(block_id)

File: csrc/memory/block_allocator.cpp
#include "block_allocator.h"

#include <glog/logging.h>

#include<vector>
namespace mllm{

BlockAllocator::BlockAllocator(int total_blocks) : total_blocks_(total_blocks), n_free_blocks_(total_blocks){
    CHECK_GT(total_blocks, 0) << "no blocks to allocate";
    free_blocks_.reserve(total_blocks);
    for(int i = total_blocks - 1;i >= 0;i --){
        free_blocks_.push_back(i);
    }
}

BlockAllocator::~BlockAllocator() {
    CHECK(free_blocks_.size() == total_blocks_) << "not all blocks are freed";
};

std::vector<int> BlockAllocator::allocate(int n_blocks){
    std::vector<int> blocks(n_blocks);
    for(int i = 0;i < n_blocks;i ++){
        CHECK_GT(n_free_blocks_, 0) << "not enough blocks";
        blocks[i] = free_blocks_[--n_free_blocks_];
    }
    return blocks;
}
    
void BlockAllocator::free(const std::vector<int>& blocks){
    for(int i = 0;i < blocks.size();i ++){
        CHECK_LT(n_free_blocks_, total_blocks_) << "free blocks overflow";
        free_blocks_[n_free_blocks_ ++] = blocks[i];
    }
}

}
